### PR TITLE
feat: hierarchical facets

### PR DESCRIPTION
**Pull request by vvo**
I have implemented the hierarchical facets as proposed in #137.

The available API is: (different from proposal)

Given records like this:

```json
{
  "name": "Flying dog easy IPA",
  "categories": {
    "lvl0": "beers",
    "lvl1": "beers > IPA",
    "lvl2": "beers > IPA > Flying dog"
  }
}
```

When using the helper:
```js
var helper = algoliasearchHelper(client, indexName, {
  hierarchicalFacets: [{
    name: 'categories', // mandatory
    attributes: ['categories.lvl0', 'categories.lvl1', 'categories.lvl2'] // mandatory, because we cannot guess what's the maximum level is (different from proposal)
    separator: '>' // optionnal, default to '>'
  }]
});

helper.toggleRefine('beers > IPA > Flying dog');

helper.on('result', function(content) {
  console.log(content.hierarchicalFacets);
  // [{
  //   'name': 'categories',
  //   'count': null,
  //   'isRefined': true,
  //   'path': null,
  //   'data': [{
  //     'name': 'beers',
  //     'path': 'beers',
  //     'count': 2,
  //     'isRefined': true,
  //     'data': [{
  //       'name': 'IPA',
  //       'path': 'beers > IPA',
  //       'count': 2,
  //       'isRefined': true,
  //       'data': [{
  //         'name': 'Flying dog',
  //         'path': 'beers > IPA > Flying dog',
  //         'count': 1,
  //         'isRefined': true,
  //         'data': null
  //       }, {
  //         'name': 'Anchor steam',
  //         'path': 'beers > IPA > Anchor steam',
  //         'count': 1,
  //         'isRefined': false,
  //         'data': null
  //       }]
  //     }]
  //   }]
  // }]
});

helper.search();
```

We always get the parent's current level values: if you refine on `beers > IPA > Flying dog`:
 - the hits will be only `Flying dog` beers
 - you will have a listing of all `Flying dog` beers and a listing of all `IPA` beers

I updated the original POC demo to use this new build (locally): it works.

Every reviewer gets one [Flying dog beer](http://flyingdogbrewery.com/beers/).

giphy-beer

### CODE REVIEW

**Comment by pixelastic** [comment]
I like it, it is quite easy to understand how this works. I have a few remarks and questions.

I think I understand why we need to repeat the previous parent categories in each `lvl` but this seems hackish. I understand that we cannot guess the priority order from the keys, but as a user I would expect a syntax with less clutter, like an `prioriies` key being an array of heading priority.

Something like:

```javascript
{
  "name": "Flying dog easy IPA",
  "categories": {
    "top_header": "beers",
    "subclass": "IPA",
    "final_page": "Flying dog"
  }
}

var helper = algoliasearchHelper(client, indexName, {
  hierarchicalFacets: [{
    name: 'categories',
    priorities: ['top_header', 'subclass', 'final_page'],
    separator: '>' // optionnal, default to '>'
  }]
});
```

Also, what is the `count` attribute in the results, and why is it `null`, shouldn't it be set to 2 ?

And finally I didn't get this comment:

> you will have a listing of all Flying dog beers and a listing of all IPA beers

If I filter on `beers > IPA > Flying dog`, won't I only have `Flying Dog` beers, so `IPA` only contains `Flying dog`?

**Comment by vvo** [comment]
> I understand that we cannot guess the priority order from the keys, but as a user I would expect a syntax with less clutter, like an priorities key being an array of heading priority.

The thing is, what happens when you refine on `subclass: oranges` and your dataset have many `oranges` categories with the parent category being different?

You won't know if the user want "smoothies > oranges" or "fruits > oranges". And since this is disjunctive faceting, you would ask for "smoothies OR oranges" which is not what you want, I want the oranges smoothies.

Thus you need the full path at each category.

> If I filter on beers > IPA > Flying dog, won't I only have Flying Dog beers, so IPA only contains Flying dog?

You will get all the IPAS beers and all the Flying dog beers, because as a navigation system this is what you want:

```txt
| beers
  | IPA
    | Flying DOG
      - first flying dog
      - second flying dog
    | Belgian IPA
    | French IPA
```

You do get the listing for all IPA beers with the Flying dog opened, so that you can easily switch, otherwise the navigation is much harder for the end user.

Basically you always have the parent level values. We could change this default and hide under an option. Or the opposite, make it the default then provide an option to deactivate.

**Comment by pixelastic** [comment]
Ok, I understand the `orange` example.

I don't understand how you managed the magic trick of getting the french and belgian IPA from a facetted search on Flying dog. Are you doing multiple queries?

**Comment by vvo** [comment]
> Are you doing multiple queries?

Yes this is the same trick than for disjunctive facet, indeed hierarchical facet is just a disjunctive facet with a specific behavior.

There's one "hitQuery" and one "parentFacetValuesQuery" query for every disjunctive facet refinement.

See https://github.com/algolia/algoliasearch-helper-js/blob/7d31913150b4e1483028ca855aa7c53b7b3cbbd5/test/spec/hierarchical-facets/simple.js#L94-L95

**Comment by pixelastic** [comment]
So what happen if you have for example 4 levels in your categories hierarchy ? Will you get facetted values for the parent category only, or facetted values for every categories excepte the top level one ? I guess this level of recursion could be defined as a config option

**Comment by vvo** [comment]
> So what happen if you have for example 4 levels in your categories hierarchy ? Will you get facetted values for the parent category only, or facetted values for every categories excepte the top level one ? I guess this level of recursion could be defined as a config option

You get only the parent level values, let's wait for the usecase to popout before we implement this option?

**Comment by vvo** [comment]
I will soon merge this, please comment if you are against it

**Comment by bobylito** [comment]
In your example, you have a path like `beers > IPA > Flying dog` and the separator is `>`. What happens with the values? Do they contain spaces?

File: src/SearchParameters/index.js [hunk 1]
```diff
@@ -726,6 +748,38 @@ SearchParameters.prototype = {
...
+   * Switch the refinement applied over a facet/value
+   * @method
+   * @param {string} facet name of the attribute used for facetting
+   * @param {value} value value used for filtering
+   * @return {SearchParameters}
+   */
+  toggleHierarchicalFacetRefinement: function toggleHierarchicalFacetRefinement(facet, value) {
+    var separator = this.getHierarchicalFacetSeparator(this.getHierarchicalFacetByName(facet));
+
+    return this.mutateMe(function merge(newInstance) {
```
Comment (bobylito) [inline src/SearchParameters/index.js]: Why didn't you use the setQueryParameters method here instead of mutateMe?

File: src/SearchResults/generate-hierarchical-tree.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+      while (level < currentHierarchicalLevel) {
+        parent = parent && find(parent.data, {isRefined: true});
+        level++;
+      }
+    }
+
+    // we found a refined parent, let's add current level data under it
+    if (parent) {
+      parent.data = sortByOrder(
```
Comment (bobylito) [inline src/SearchResults/generate-hierarchical-tree.js]: Pfoushh this `sortByOrder` just made my day ;)

File: src/SearchResults/index.js [hunk 3]
```diff
@@ -218,56 +238,76 @@ function SearchResults(state, algoliaResponse) {
...
+          data: extend({}, dataFromMainRequest, facetResults),
+          exhaustive: result.exhaustiveFacetsCount
+        };
+        assignFacetStats(this.disjunctiveFacets[position], result.facets_stats, dfacet);
+
+        if (state.disjunctiveFacetsRefinements[dfacet]) {
+          forEach(state.disjunctiveFacetsRefinements[dfacet], function(refinementValue) {
+            // add the disjunctive refinements if it is no more retrieved
+            if (!this.disjunctiveFacets[position].data[refinementValue] &&
+              indexOf(state.disjunctiveFacetsRefinements[dfacet], refinementValue) > -1) {
```
Comment (bobylito) [inline src/SearchResults/index.js]: Isn't there an indentation problem here?

**Comment by bobylito** [comment]
Looks good so far. I raised some stuff some questions though ;)

**Comment by vvo** [comment]
> In your example, you have a path like beers > IPA > Flying dog and the separator is >. What happens with the values? Do they contain spaces?

No it will not contain spaces, values are trimmed.

We had a little talk with @bobylito, I did some magic and I did not explain why and how.

So to be more precise, if for a product you have `categories: {lvl0: 'beer', lvl1: 'beer > IPA'}` or `categories: {lvl0: 'beer', lvl1: 'beer>IPA'}` (notice, no space around `>`): Then the hierarchical result will be the same.

I do `trim()` after splitting on the separator `>`. This means that if you have spaces or not it does not matter.

The question is, should we handle this for ou users or not?

Pros:
- The user doesn't need to reformat his data or search the documentation for the separator whether he has spaces or no spaces around the separator

Cons:
- You cannot say that your categories separator is a ` ` (space character). Nor can it be a <TAB>)

Even if I do not like magic in programming, here it does brings something useful since I do not want to have support requests like: "I have weird spaces in my <a href=""></a> of the hierarchical menu, why?".

Also I should add that I originally did the examples and API proposal with a space around the separator because the POC I did with some client data was using spaces.

**Comment by maxiloc** [comment]
@vvo Yes I would let the trim. But do you format it with a space for display ?

**Comment by vvo** [comment]
after discussions with @maxiloc and @bobylito this need some more work, Ill update asap

**Comment by vvo** [comment]
This PR is good to go.

tumblr_mrhfdlmusa1relg8bo1_250

**Comment by bobylito** [comment]
MERGEEE
